test/compatibility/dogstatsd_datagram_compatibility_test.rb

```diff
@@ -17,56 +17,89 @@ def teardown
+      assert_equal_datagrams { |client| client.set('unique', '1234', tags: ['foo', 'bar']) }
+      assert_equal_datagrams { |client| client.set('unique', '1234', tags: { foo: 'bar' }) }
+      assert_equal_datagrams { |client| client.set('unique', '1234', sample_rate: 0.1, tags: ['quc']) }
     end
 
     def test_histogram_compatibility
-      assert_equal_datagrams_emitted { |client| client.histogram('sample', 12.44) }
-      assert_equal_datagrams_emitted { |client| client.histogram('sample', 12.44, sample_rate: 0.1) }
-      assert_equal_datagrams_emitted { |client| client.histogram('sample', 12.44, tags: ['foo', 'bar']) }
-      assert_equal_datagrams_emitted { |client| client.histogram('sample', 12.44, tags: { foo: 'bar' }) }
-      assert_equal_datagrams_emitted { |client| client.histogram('sample', 12.44, sample_rate: 0.1, tags: ['quc']) }
+      assert_equal_datagrams { |client| client.histogram('sample', 12.44) }
+      assert_equal_datagrams { |client| client.histogram('sample', 12.44, sample_rate: 0.1) }
+      assert_equal_datagrams { |client| client.histogram('sample', 12.44, tags: ['foo', 'bar']) }
+      assert_equal_datagrams { |client| client.histogram('sample', 12.44, tags: { foo: 'bar' }) }
+      assert_equal_datagrams { |client| client.histogram('sample', 12.44, sample_rate: 0.1, tags: ['quc']) }
     end
 
     def test_distribution_compatibility
-      assert_equal_datagrams_emitted { |client| client.distribution('sample', 12.44) }
-      assert_equal_datagrams_emitted { |client| client.distribution('sample', 12.44, sample_rate: 0.1) }
-      assert_equal_datagrams_emitted { |client| client.distribution('sample', 12.44, tags: ['foo', 'bar']) }
-      assert_equal_datagrams_emitted { |client| client.distribution('sample', 12.44, tags: { foo: 'bar' }) }
-      assert_equal_datagrams_emitted { |client| client.distribution('sample', 12.44, sample_rate: 0.1, tags: ['quc']) }
+      assert_equal_datagrams { |client| client.distribution('sample', 12.44) }
+      assert_equal_datagrams { |client| client.distribution('sample', 12.44, sample_rate: 0.1) }
+      assert_equal_datagrams { |client| client.distribution('sample', 12.44, tags: ['foo', 'bar']) }
+      assert_equal_datagrams { |client| client.distribution('sample', 12.44, tags: { foo: 'bar' }) }
+      assert_equal_datagrams { |client| client.distribution('sample', 12.44, sample_rate: 0.1, tags: ['quc']) }
+    end
+
+    def test_distribution_with_block_compatibility
+      Process.stubs(:clock_gettime).with(Process::CLOCK_MONOTONIC).returns(12.1)
+      assert_equal_datagrams do |client|
+        return_value = client.distribution('timing', tags: ['foo'], sample_rate: 0.1) { 'foo' }
+        assert_equal 'foo', return_value
+      end
     end
 
     private
 
-    def with_legacy_client
+    MODES = [:normal, :with_prefix, :with_default_tags]
+
+    def assert_equal_datagrams(&block)
+      MODES.each do |mode|
```

Comment:
So you edited your PR.

I think dynamically defining a test for each mode would be preferable:

```ruby
MODES.each do |mode|
  define_method("test_blah_in_#{mode}_mode") do
    # test
  end
end
```

This way if you break one of the mode it's much more clear.

Reply:
I've tried this, but the code gets quite painful because we have to modify the StatsD singleton when we switch modes, and put it back in its original state.

Instead, I've made sure that if a datagram is not equal, the mode is included in the assertion method.